homeassistant/components/zwave_js/humidifier.py

```diff
@@ -0,0 +1,188 @@
+from __future__ import annotations
+
+from typing import cast
+
+from zwave_js_server.client import Client as ZwaveClient
+from zwave_js_server.const import CommandClass
+from zwave_js_server.const.command_class.humidity_control import (
+    HUMIDITY_CONTROL_MODE_PROPERTY,
+    HUMIDITY_CONTROL_MODE_SETPOINT_MAP,
+    HUMIDITY_CONTROL_OPERATING_STATE_PROPERTY,
+    HUMIDITY_CONTROL_SETPOINT_PROPERTY,
+    HUMIDITY_CONTROL_SUPPORTED_SETPOINT_TYPES_PROPERTY,
+    HumidityControlMode,
+    HumidityControlSetpointType,
+)
+from zwave_js_server.model.value import Value as ZwaveValue
+
+from homeassistant.components.humidifier import HumidifierDeviceClass, HumidifierEntity
+from homeassistant.components.humidifier.const import (
+    DEFAULT_MAX_HUMIDITY,
+    DEFAULT_MIN_HUMIDITY,
+    DOMAIN as HUMIDIFIER_DOMAIN,
+    SUPPORT_MODES,
+    SUPPORT_OPERATING_STATE,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DATA_CLIENT, DOMAIN
+from .discovery import ZwaveDiscoveryInfo
+from .entity import ZWaveBaseEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Z-Wave humidifier from config entry."""
+    client: ZwaveClient = hass.data[DOMAIN][config_entry.entry_id][DATA_CLIENT]
+
+    @callback
+    def async_add_humidifier(info: ZwaveDiscoveryInfo) -> None:
+        """Add Z-Wave Humidifier."""
+        entities: list[ZWaveBaseEntity] = []
+
+        entities.append(ZWaveHumidifier(config_entry, client, info))
+
+        async_add_entities(entities)
+
+    config_entry.async_on_unload(
+        async_dispatcher_connect(
+            hass,
+            f"{DOMAIN}_{config_entry.entry_id}_add_{HUMIDIFIER_DOMAIN}",
+            async_add_humidifier,
+        )
+    )
+
+
+class ZWaveHumidifier(ZWaveBaseEntity, HumidifierEntity):
+    """Representation of a Z-Wave Humidifier."""
+
+    def __init__(
+        self,
+        config_entry: ConfigEntry,
+        client: ZwaveClient,
+        info: ZwaveDiscoveryInfo,
+    ) -> None:
+        """Initialize thermostat."""
+        super().__init__(config_entry, client, info)
+
+        self._attr_supported_features = SUPPORT_MODES | SUPPORT_OPERATING_STATE
+        self._attr_available_modes = []
+
+        self._current_mode = self.get_zwave_value(
+            HUMIDITY_CONTROL_MODE_PROPERTY,
+            command_class=CommandClass.HUMIDITY_CONTROL_MODE,
+        )
+
+        if self._current_mode:
+            if HumidityControlMode.HUMIDIFY in self._current_mode.metadata.states:
+                self._attr_device_class = HumidifierDeviceClass.HUMIDIFIER
+            else:
+                self._attr_device_class = HumidifierDeviceClass.DEHUMIDIFIER
```

Comment:
maybe submit a PR for the platform to have a new device class that means both?

Reply:
Opened https://github.com/home-assistant/core/pull/65858

Reply:
To avoid blocking this PR I'd make the assumption that we can set the device class to humidifier even though the device does both?

Reply:
After reading through https://github.com/home-assistant/architecture/issues/310 I found https://github.com/home-assistant/architecture/issues/310#issuecomment-719752004 which seems to point towards splitting the humidifying and de-humidifying capabilities into two entities. This would mean the z-wave `Humidity control mode` would need to map to the on/off state of each device in this way:

| Humidifier entity | De-humidifier entity | Z-Wave Humidity Control Mode |
|----|----|----|
| OFF | OFF | OFF |
| ON | OFF | Humidify |
| OFF | ON | De-humidify |
| ON | ON | Auto |

Thoughts on this approach? Still seems weird to me because two entities need to know about each other's state to control the underlying z-wave value correctly.

Reply:
I'd suggest to simplify and do last service call wins. Ie, not have any interdependence between entities.

Reply:
I'll move this back to draft and make the changes to split these in two entities

Reply:
I'm thinking about this a bit more, there's no need for two entities to have a dependency on each other. This can be mediated through the underlying zwave value which would be shared between both entities regardless.

On the humidifier side we will have:
```python
def turn_on():
  if current_mode == OFF:
    current_mode = HUMIDIFY
  elif current_mode == DEHUMIDIFY:  # turn on humidifier from "de-humidify"-only by setting the mode to auto
    current_mode = AUTO

def turn_off():
  if current_mode = AUTO: # turn off humidifier from "auto" by setting the mode to de-humidify only
    current_mode = DEHUMIDIFY
  elif current_mode = HUMIDIFY:
    current_mode = OFF
```

On the de-humidifier side:
```python
def turn_on():
  if current_mode == OFF:
    current_mode = DEHUMIDIFY
  elif current_mode == HUMIDIFY: # turn on de-humidifier from "humidify"-only by setting the mode to auto
    current_mode = AUTO

def turn_off():
  if current_mode = AUTO: # turn off de-humidifier from "auto" by setting the mode to humidify only
    current_mode = HUMIDIFY
  elif current_mode = DEHUMIDIFY:
    current_mode = OFF
```